gnucash/gnucash-cli.cpp

```diff
@@ -126,11 +144,117 @@ may be specified to describe some saved options.\n"
+            std::cout << "Running script from " << *args->script << "...";
+        scm_c_primitive_load (args->script->c_str());
+        if (args->verbose)
+            std::cout << "success!" << std::endl;
+    }
+    if (args->interactive)
+    {
+        if (args->verbose)
+            std::cout << "Starting CLI... " << std::endl;
+        scm_c_eval_string ("(start-repl)");
+    }
+    if (gnc_current_session_exist())
+    {
+        if (args->verbose)
+            std::cout << "Warning: session still open. Clearing session..." << std::endl;
+        gnc_clear_current_session ();
+    }
+}
+
+static void load_file (std::string m_file_to_load, bool m_open_readwrite, bool m_verbose)
+{
+    if (m_verbose)
+        std::cout << "\n\nLoading " << m_file_to_load
+                  << (m_open_readwrite ? " (r/w)" : " (readonly)");
+    auto session = gnc_get_current_session();
+    auto mode = m_open_readwrite ? SESSION_READ_ONLY : SESSION_NORMAL_OPEN;
+    while (true)
+    {
+        qof_session_begin (session, m_file_to_load.c_str(), mode);
+        auto io_err = qof_session_get_error (session);
+        switch (io_err)
+        {
+        case ERR_BACKEND_NO_ERR:
+            qof_session_load (session, NULL);
+            if (m_verbose)
+                std::cout << "... done!\n\n";
+            return;
+        case ERR_BACKEND_LOCKED:
+        case ERR_BACKEND_READONLY:
+            if (m_verbose)
+                std::cout << " (forced readonly)";
+            mode = SESSION_READ_ONLY;
+            break;
+        default:
+            if (m_verbose)
+                std::cout << "... unknown error. Abort." << std::endl;
+            exit (1);
+        }
+    }
+}
+
 int
 Gnucash::GnucashCli::start ([[maybe_unused]] int argc, [[maybe_unused]] char **argv)
 {
     Gnucash::CoreApp::start();
 
+    if (m_interactive || m_script)
+    {
+        if (m_open_readwrite && !m_file_to_load)
+        {
+            std::cerr << "missing datafile to be open read/write!" << std::endl;
+            return 1;
+        }
+        if (m_script && (!std::filesystem::exists (*m_script) ||
+                         std::filesystem::is_directory (*m_script)))
+        {
+            std::cerr << "invalid script " << *m_script << std::endl;
+            return 1;
+        }
+        gnc_prefs_init ();
+        cli_struct args = { m_script, m_interactive, m_verbose };
+        if (!m_language || *m_language == "guile")
+        {
+            if (m_file_to_load)
+                load_file (*m_file_to_load, m_open_readwrite, m_verbose);
+            scm_boot_guile (0, nullptr, run_guile_cli, &args);
+            return 0;
+        }
+        else if (*m_language == "python")
+        {
+            std::cerr << "don't know how to launch python" << std::endl;
+            return 1;
+        }
+        else
+        {
+            std::cerr << "Valid languages are 'python' or 'guile'" << std::endl;
+            return 1;
+        }
+    }
```

Comment:
IMO this needs a gnc_shutdown_cli here to prevent the code from continuing to parse for command line options after an interactive session ends. That's more obvious when this bit is moved to gnucash-commands.cpp as you'll find most of the other commands end like that.

Reply:
i think it's done.